### ISSUE TITLE
Core Contributor Metrics with GrimoireLab

### ISSUE THREAD

**Issue by cassiezamparini** [openedx/wg-coordination#139]
See proposal: https://openedx.atlassian.net/wiki/spaces/COMM/pages/4522278913/Core+Contributor+Metrics+with+GrimoireLab

**Comment by jalondonot** [comment]
@pomegranited, during the [last meeting of the Coordination Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/4643323905/2024-11-26+CC+Working+Group+Meeting+Notes), we agreed that creating a work plan for each CC Enhancement initiative would be beneficial. We understand that progress may be challenging due to the Christmas and New Year's holidays. However, when you resume work in January (I guess), we would greatly appreciate it if you could share the next steps to advance your initiative. This will help engage other members of the Open edX community who may be interested in supporting your efforts. Please let us know once you have a timeline in mind so that we can assist you in any way possible. Thanks and Merry Christmas!

**Comment by pomegranited** [comment]
Hi @jalondonot thank you for the greetings and nudge about this work!

I'm focusing my CC time in January on some Aspects issues, trying get onboard as a Superset contributor. So I wasn't going to start this task until February. But I'm happy to discuss any ideas you have in the meantime?

**Comment by jalondonot** [comment]
Sounds good to me, @pomegranited! I've got a pretty busy start to the year too so we can move this along in late January/early February. Good luck with your January backlog!

**Comment by antoviaque** [comment]
@pomegranited Are you able to put some work into this this month? Note that if your other core contributor duties keep you too busy, @pkulkark has shown interest for the Open edX bitberg.io board, so maybe you can work out something together :)

**Comment by pomegranited** [comment]
Hi @antoviaque , and my apologies for not communicating here @jalondonot !

@pkulkark has started work on this feature, and is communicating with the contributors for these initiatives to find out whether the existing data being tracked around the [Github PRs Efficiency](https://openedx.biterg.io/app/dashboards#/view/9663d5a0-e1dc-11e8-8aac-ef7fd4d8cbad?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-90d,to:now))&_a=(description:'Efficiency%20panel%20for%20GitHub%20Pull%20Requests%20by%20Bitergia',filters:!(('$state':(store:appState),meta:(alias:Bots,disabled:!f,index:github_issues,key:author_bot,negate:!t,params:(query:!t,type:phrase),type:phrase,value:true),query:(match:(author_bot:(query:!t,type:phrase))))),fullScreenMode:!f,options:(darkTheme:!f,hidePanelTitles:!f,useMargins:!t),query:(language:lucene,query:''),timeRestore:!f,title:'GitHub%20Pull%20Requests%20Efficiency',viewMode:view)) (REI and lead time) could be used to measure improvements in the OSPR process, and what additional metrics may be helpful.

* https://github.com/openedx/wg-coordination/issues/134
* https://github.com/openedx/wg-coordination/issues/135
* https://github.com/openedx/wg-coordination/issues/140
* https://github.com/openedx/wg-coordination/issues/136

**Comment by antoviaque** [comment]
@pomegranited Thanks for the update!

@pkulkark How is that going on your end? Any blockers?

**Comment by antoviaque** [comment]
@pkulkark Ping on my question ^

**Comment by pkulkark** [comment]
@antoviaque Apologies for the delay in response. This got buried in my email backlog. Here's the update:

I posted on each of the other proposals mentioned earlier and these are the status:
1. OSPR Management: The proposal work was put on the back burner due to budget usage but will be looking into it adding the bitergia data to it soon.
2. Single Person Responsible: This proposal has been updated to include the bitergia data.
3. Improve Review Process: This was merged with the OSPR management proposal.
4. Partner as Maintainers: The available data from Bitergia could be used partially here but would benefit from more fine-grained data like in cases of one org contributing on behalf of another. But implementing a way of tracking that info would add more overhead and likely not accepted upstream.

**Comment by antoviaque** [comment]
@pkulkark Thanks for the update!

> OSPR Management: The proposal work was put on the back burner due to budget usage but will be looking into it adding the bitergia data to it soon.

OK. When do you think you will have the time & budget for it?

> Single Person Responsible: This proposal has been updated to include the bitergia data.
> Improve Review Process: This was merged with the OSPR management proposal.

For reference, could you explain how this was integrated with those proposals?

> Partner as Maintainers: The available data from Bitergia could be used partially here but would benefit from more fine-grained data like in cases of one org contributing on behalf of another. But implementing a way of tracking that info would add more overhead and likely not accepted upstream.

What wouldn't be accepted upstream exactly? Note that we aren't completely sure what will end up being accepted there, but whatever ends up being the criteria will be quite important to track, so there should be good arguments to get whatever is needed to track it to be merged.

**Comment by pkulkark** [comment]
@antoviaque

> When do you think you will have the time & budget for it?

I'm not the owner of that proposal, @itsjeyd is. I will defer the question to him.

> For reference, could you explain how this was integrated with those proposals?

Sure. For the Single Person Responsible, the bitergia data being tracked around the [Github PRs Efficiency](https://openedx.biterg.io/app/dashboards#/view/9663d5a0-e1dc-11e8-8aac-ef7fd4d8cbad?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-90d,to:now))&_a=(description:'Efficiency%20panel%20for%20GitHub%20Pull%20Requests%20by%20Bitergia',filters:!(('$state':(store:appState),meta:(alias:Bots,disabled:!f,index:github_issues,key:author_bot,negate:!t,params:(query:!t,type:phrase),type:phrase,value:true),query:(match:(author_bot:(query:!t,type:phrase))))),fullScreenMode:!f,options:(darkTheme:!f,hidePanelTitles:!f,useMargins:!t),query:(language:lucene,query:''),timeRestore:!f,title:'GitHub%20Pull%20Requests%20Efficiency',viewMode:view)) (REI and lead time) will be used to measure improvements done.

> What wouldn't be accepted upstream exactly? Note that we aren't completely sure what will end up being accepted there, but whatever ends up being the criteria will be quite important to track, so there should be good arguments to get whatever is needed to track it to be merged.

For tracking the work that is done by one org on behalf of other organizations, the contributions would need to be tagged with something like "contributed for <company>"  or "attribute to <org>". Bitergia could then be used to track it. It is likely to be confusing, though, given the already large number of tags used and the nuances involved.

**Comment by itsjeyd** [comment]
>> When do you think you will have the time & budget for it?
>
> I'm not the owner of that proposal, @itsjeyd is. I will defer the question to him.

@pkulkark @antoviaque With the current limits on Contribution hours I don't have a lot of time to spend on anything but OSPR management work itself. But I should be able to follow up on how this proposal overlaps with @xitij2000's and mine on improving OSPR management this month.

**Comment by antoviaque** [comment]
@pkulkark @itsjeyd OK, thanks!

> For tracking the work that is done by one org on behalf of other organizations, the contributions would need to be tagged with something like "contributed for " or "attribute to ". Bitergia could then be used to track it. It is likely to be confusing, though, given the already large number of tags used and the nuances involved.

@pkulkark Ah, I see. For the "partners as maintainers", the tracking looks likely to be in terms of resources dedicated to contributions (ie full time equivalents or number of hours). So it would likely be a separate metric. The one you mention would be a good one to have too, but it's true that it is likely tricky to calculate in practice, at least in terms of accurate organizational attribution.

**Comment by itsjeyd** [comment]
Just a quick update, I ran out of contributions hours again this month before getting to [this](https://github.com/openedx/wg-coordination/issues/139#issuecomment-2804108689), but I will prioritize it for early May to make sure it gets done.

**Comment by antoviaque** [comment]
@itsjeyd This is still on your radar? When do you expect to be able to block some contribution hours for it?

**Comment by itsjeyd** [comment]
@antoviaque Yep, I still have some contribution hours blocked for it and should be able to get to it this week.

**Comment by itsjeyd** [comment]
@pkulkark @antoviaque I checked out the Bitergia dashboard and after playing around with it for a bit I added it as a preferred option for evaluating the impact of changes to OSPR management to the proposal from https://github.com/openedx/wg-coordination/issues/134.

So that part is done.

In terms of adding more metrics to the dashboard, that probably won't be necessary; though my thinking might change depending on how the question that I posted [here](https://tasks.opencraft.com/browse/STAR-3870?focusedCommentId=310910&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-310910) is answered.

**Comment by pomegranited** [comment]
Thank you for evaluating and using these dashboards @itsjeyd ! CC @pkulkark 

Answering your question here for transparency:

> One question that I had was, what's the difference between Time to Merge and Lead Time? Conceptually, I mean. I do realize that the dashboard shows different metrics for them (median vs. average) 😅

Yeah, they're only subtly different.. "Time to Merge" only includes PRs that were merged (not closed/rejected).

More info: https://bitergia.com/lead-time-metric-issues-code-changes/

**Comment by itsjeyd** [comment]
Thanks @pomegranited! What you're saying about how these metrics differ conceptually makes sense to me.

The documentation that you linked to confused me, though. Maybe I misread it, but some of the info in there seemed contradictory to me:

> The Lead Time is defined as the time, expressed in days, between initiating and completing a production process. ...
> 
> For Issues, this metric measures the time a software development team takes to resolve and close issues (also referred to as “Time to Issue Resolution” or “Time to Close”). For code changes (such as Pull Requests), it measures the time to complete review processes (also referred to as “Change Requests Duration” or “Time to Merge”) from when they were opened until they were merged, accepted, or closed.
> 
> This metric can be calculated using the following formula:
> ```
> Lead Time = Time when the item was closed – Time when it was opened
> ```
> ...
> 
> ### Time to Close for Issues
> 
> ... this metric is a specific implementation of the Lead Time metric, focused on issues, reviews, or tickets. ... the time to close is the total amount of time, usually expressed in days, that passes between the creation and closing of an operation, such as an issue, review, or support ticket. ...
> 
> This metric can be calculated as follows:
> ```
> Time to Close (days) = Time when the item was closed – Time when it was opened
> ```
> ...

So far so good. But then it says:

> ### Change Requests Duration
>
> This metric is also a specific implementation of the Lead Time metric, this time for Change Requests. The difference with the Time to Close metric is that **this one considers the time between the moment the Change Request was sent to the moment it is _accepted or merged_**, considering all the steps in the development process, including coding, testing, reviewing, and merging. 
> 
> ... This metric can be calculated as follows:
> 
> Change Request Duration (days) = **Time when the item was _merged, accepted, or closed_** – Time when the item was opened.
> 
> Similar to what we represented for the Time to Close for Issues, we usually represent this metric as “Time to Merge”. ...

And then there's a screenshot of the same metrics that the [Open edX Bitergia dashboard](https://openedx.biterg.io/app/dashboards#/view/9663d5a0-e1dc-11e8-8aac-ef7fd4d8cbad?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-12M,to:now))&_a=(description:'Efficiency%20panel%20for%20GitHub%20Pull%20Requests%20by%20Bitergia',filters:!(('$state':(store:appState),meta:(alias:Bots,disabled:!f,index:github_issues,key:author_bot,negate:!t,params:(query:!t,type:phrase),type:phrase,value:true),query:(match:(author_bot:(query:!t,type:phrase))))),fullScreenMode:!f,options:(darkTheme:!f,hidePanelTitles:!f,useMargins:!t),query:(language:lucene,query:''),timeRestore:!f,title:'GitHub%20Pull%20Requests%20Efficiency',viewMode:view)) displays for PRs (Time to Merge and Lead Time), and the caption refers to both metrics as "time/days to merge":

![screenshot-bitergia-time-to-merge-lead-time](https://github.com/user-attachments/assets/fe8d77f0-ff22-466f-ab4f-0b6ee6738e06)

> The visualization on the right represents the **average days to merge** over time, as in the previous metric. Also, we are representing on the left the **median time to merge** for the whole period of time ....

Let me know what you think. Did you have a chance to look at how these metrics are actually implemented to confirm that time-to-merge doesn't take into account closed PRs, while lead time for PRs does?

CC @pkulkark

**Comment by pomegranited** [comment]
@itsjeyd 

> Did you have a chance to look at how these metrics are actually implemented to confirm that time-to-merge doesn't take into account closed PRs, while lead time for PRs does?

Sorry no, I didn't have time to dig into the data that feeds these metrics. It's not an easy task, unfortunately!

> Change Request Duration (days) = Time when the item was merged, accepted, or closed – Time when the item was opened.

Ugh.. that is confusing!

That indicates that "Time to Merge" and "Lead time" are using the same dataset, so their only difference is the "median" vs "average"? If that's the case, then our median "Time to Merge" is preposterously low at .2 days, while our average "Lead Time" trend is a more reasonable 10-20 days. I guess the data could be that skewed, but it seems unlikely. :(

**Comment by itsjeyd** [comment]
@pomegranited The screenshot above is from the Bitergia documentation, it doesn't represent Open edX data. That said the situation doesn't seem to be all that different for Open edX. Looking at the [last 90 days](https://openedx.biterg.io/app/dashboards#/view/9663d5a0-e1dc-11e8-8aac-ef7fd4d8cbad?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-90d,to:now))&_a=(description:'Efficiency%20panel%20for%20GitHub%20Pull%20Requests%20by%20Bitergia',filters:!(('$state':(store:appState),meta:(alias:Bots,disabled:!f,index:github_issues,key:author_bot,negate:!t,params:(query:!t,type:phrase),type:phrase,value:true),query:(match:(author_bot:(query:!t,type:phrase))))),fullScreenMode:!f,options:(darkTheme:!f,hidePanelTitles:!f,useMargins:!t),query:(language:lucene,query:''),timeRestore:!f,title:'GitHub%20Pull%20Requests%20Efficiency',viewMode:view)), median time to merge was 1.9 days while average lead time was hovering around 25-26 days.

> Sorry no, I didn't have time to dig into the data that feeds these metrics. It's not an easy task, unfortunately!

No worries. Maybe reaching out to Bitergia devs and raising the question with them would be an option?

**Comment by pomegranited** [comment]
@itsjeyd 
> Maybe reaching out to Bitergia devs and raising the question with them would be an option?

Oh sure thing! I couldn't find a forum or Slack channel anywhere, so contacted via their form: https://bitergia.com/contact/ 
Will keep you posted :)

**Comment by itsjeyd** [comment]
Great, thanks! :)

**Comment by pomegranited** [comment]
@itsjeyd Thank you for encouraging me to reach out to bitergia -- they responded so quickly! Here's what they sent, does this explain it sufficiently?

These line visualizations are not 100% integrated with the dashboards, so there's indeed a difference:

The line visualization shows PRs **closed** in the timeframe whereas the gauge filters by **creation** date. Therefore the gauge captures a smaller sample of shorted-lived PRs than the line visualization, which includes older PRs with longer durations that push the average up).

I've calculated the average Time to Merge (as the gauge would show it) and is around 5 days. You could simulate this by applying following filter:

![Image](https://github.com/user-attachments/assets/4b908ef0-83b8-4d2b-86bd-851fbbf439cd)

The [distribution of lead time for (closed) PRs](https://openedx.biterg.io/app/visualize#/edit/2cd4e170-3b1b-11f0-bf7e-eb3d898689ef) (created within the last 90 days) has the shape of a long tail (or a deep, narrow funnel depending on your reference axis):

![Image](https://github.com/user-attachments/assets/a8daa6ac-8da1-427f-9409-d1f1875af04a)

Most of the PRs are closed very soon. Yet a small few stay for very long. From ~1700 PRs,

* \>650 are closed almost immediately
* 50% are closed within 1.3 days
* The other 50% spreads over a long tail of durations. This tail is very thin, but it si consistent (there are PRs for almost all possible values).
* The numbers of this half of the population push the average up to 5 days.

**Comment by antoviaque** [comment]
@pomegranited Super interesting! Thank you for working on this and reviewing the numbers and definitions 👍 

Are those numbers for all PRs on the openedx repository, or just OSPRs? I assume the PRs that are merged quickly largely come from teams who can review their own work, like edX, Axim, core contributors and blended projects? The slower PRs to merge being contributions which don't have access to this "review facilitation"?

**Comment by pomegranited** [comment]
@antoviaque Yes, these numbers are for all PRs. We can [filter out PRs from authors in the edX or Axim org](https://openedx.biterg.io/app/dashboards#/view/9663d5a0-e1dc-11e8-8aac-ef7fd4d8cbad?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-12M,to:now))&_a=(description:'Efficiency%20panel%20for%20GitHub%20Pull%20Requests%20by%20Bitergia',filters:!(('$state':(store:appState),meta:(alias:Bots,disabled:!f,index:github_issues,key:author_bot,negate:!t,params:(query:!t),type:phrase),query:(match:(author_bot:(query:!t,type:phrase)))),('$state':(store:appState),meta:(alias:!n,disabled:!f,index:github_issues,key:author_multi_org_names,negate:!t,params:!('edX,%20Inc.','The%20Center%20for%20Reimagining%20Learning'),type:phrases,value:'edX,%20Inc.,%20The%20Center%20for%20Reimagining%20Learning'),query:(bool:(minimum_should_match:1,should:!((match_phrase:(author_multi_org_names:'edX,%20Inc.')),(match_phrase:(author_multi_org_names:'The%20Center%20for%20Reimagining%20Learning'))))))),fullScreenMode:!f,options:(darkTheme:!f,hidePanelTitles:!f,useMargins:!t),query:(language:lucene,query:''),timeRestore:!f,title:'GitHub%20Pull%20Requests%20Efficiency',viewMode:view)), but interestingly, there's still an order of magnitude difference between median vs mean, even in the last 12 months.

**Comment by antoviaque** [comment]
@pomegranited OK, interesting. This is just based on the github org of the contributor, though, right? So it would still include PRs from blended projects or sub-hiring like Edly's teams working for edX, no? If so, it would be great to have a way to filter those out too, as they probably inflate the proportion of PRs being merged quickly.

**Comment by pomegranited** [comment]
@antoviaque  

>  This is just based on the github org of the contributor, though, right? 

Yes. I'm also not sure when this org membership is recorded or taken into account -- e.g. what about people who were part of edX but now aren't?

> So it would still include PRs from blended projects or sub-hiring like Edly's teams working for edX, no?

Yes. If people are good about including `[FC-XXXX`] or `[BD-XXXX]` in the titles of their PRs, we can [exclude them with filters like this](https://openedx.biterg.io/app/dashboards#/view/9663d5a0-e1dc-11e8-8aac-ef7fd4d8cbad?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-12M,to:now))&_a=(description:'Efficiency%20panel%20for%20GitHub%20Pull%20Requests%20by%20Bitergia',filters:!(('$state':(store:appState),meta:(alias:Bots,disabled:!f,index:github_issues,key:author_bot,negate:!t,params:(query:!t),type:phrase),query:(match:(author_bot:(query:!t,type:phrase)))),('$state':(store:appState),meta:(alias:!n,disabled:!f,index:github_issues,key:author_multi_org_names,negate:!t,params:!('edX,%20Inc.','The%20Center%20for%20Reimagining%20Learning'),type:phrases,value:'edX,%20Inc.,%20The%20Center%20for%20Reimagining%20Learning'),query:(bool:(minimum_should_match:1,should:!((match_phrase:(author_multi_org_names:'edX,%20Inc.')),(match_phrase:(author_multi_org_names:'The%20Center%20for%20Reimagining%20Learning')))))),('$state':(store:appState),meta:(alias:!n,disabled:!f,index:github_issues,key:title,negate:!t,params:(query:%5BBD),type:phrase),query:(match_phrase:(title:%5BBD))),('$state':(store:appState),meta:(alias:!n,disabled:!f,index:github_issues,key:title,negate:!t,params:(query:%5BFC),type:phrase),query:(match_phrase:(title:%5BFC)))),fullScreenMode:!f,options:(darkTheme:!f,hidePanelTitles:!f,useMargins:!t),query:(language:lucene,query:''),timeRestore:!f,title:'GitHub%20Pull%20Requests%20Efficiency',viewMode:view)). However, that filter doesn't have much effect on the mean/median gap either, so it doesn't much help.

**Comment by itsjeyd** [comment]
@pomegranited Thanks a lot for reporting back with the details from Bitergia. Happy to hear that they were quick to respond! I'm a bit short on time this week so can't take a closer look right now, but it's on my list for next week.

**Comment by itsjeyd** [comment]
@pomegranited I had a closer look at Bitergia's response. This part in particular does seem relevant for understanding the different metrics better in general:

> The line visualization shows PRs **closed** in the timeframe whereas the gauge filters by **creation** date. Therefore the gauge captures a smaller sample of shorted-lived PRs than the line visualization, which includes older PRs with longer durations that push the average up).

I'm not sure, though, if their response answers the earlier question of whether both Lead Time and Time to Merge take into account merged **and** closed/rejected PRs?

(Maybe it does... It's been a long day 😅)

**Comment by pomegranited** [comment]
@itsjeyd 

> I'm not sure, though, if their response answers the earlier question of whether both Lead Time and Time to Merge take into account merged and closed/rejected PRs?

Oh you're right.. I've asked Bitergia support, will let you know :)

**Comment by pomegranited** [comment]
@itsjeyd Bitergia says:

> Thank you very much for your question. It has helped us find some bugs in our captions and help texts.
>
> Both metrics include all **closed** PRs: [https://chaoss.community/kb/metric-time-to-close/](https://chaoss.community/kb/metric-time-to-close/)

**Comment by itsjeyd** [comment]
@pomegranited Thanks for bearing with me here, and chatting to Bitergia to clarify all those little ambiguities. I don't have any additional questions at the moment.

Going forward, it might be helpful to update the Open edX dashboard to include charts incorporating the [filters you came up with](https://github.com/openedx/wg-coordination/issues/139#issuecomment-2931512090) for excluding `[FC-XXXX]` PRs. @pkulkark Maybe we can chat about that a bit more in late July/early August.
